shared/Microsoft.AspNetCore.Hosting.WebHostBuilderFactory.Sources/WebHostFactoryResolver.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Hosting.WebHostBuilderFactory
+{
+    internal class WebHostFactoryResolver
+    {
+        public static readonly string CreateWebHostBuilder = nameof(CreateWebHostBuilder);
+        public static readonly string BuildWebHost = nameof(BuildWebHost);
+
+        public static FactoryResolutionResult ResolveWebHostBuilderFactory(Assembly assembly)
+        {
+            var programType = assembly?.EntryPoint?.DeclaringType;
+            if (programType == null)
+            {
+                return FactoryResolutionResult.NoEntryPoint();
+            }
+
+            var factory = programType?.GetTypeInfo().GetDeclaredMethod(CreateWebHostBuilder);
+            if (factory == null)
+            {
+                return FactoryResolutionResult.NoCreateWebHostBuilder(programType);
+            }
+
+            return FactoryResolutionResult.Succeded(args => (IWebHostBuilder)factory.Invoke(null, new object[] { args }), programType);
+        }
+
+        public static FactoryResolutionResult ResolveWebHostFactory(Assembly assembly)
```

Comment:
No tests for this.

Reply:
I didn't want to create a test for this in Hosting as it would require another project. My strategy was to have the tests on the MVC package (that consumes this) and on EF that already has tests for this on the dotnet tool. I can add all the tests here if you prefer, but it would be very verbose IMO and it's already going to be tested at higher levels of the stack.

Reply:
Add the project, that's not significant overhead. It's better for the tests to live adjacent to the code.